src/Storages/Statistics/ConditionEstimator.cpp

```diff
@@ -112,26 +162,40 @@ Float64 ConditionEstimator::estimateSelectivity(const RPNBuilderTreeNode & node)
     auto [op, val] = extractBinaryOp(node, col);
     if (op == "equals")
     {
-        if (val < - threshold || val > threshold)
-            return default_normal_cond_factor;
-        else
-            return default_good_cond_factor;
+        if (dummy)
+        {
+            if (val < - threshold || val > threshold)
+                return default_normal_cond_factor * total_rows;
+            else
+                return default_good_cond_factor * total_rows;
+        }
+        return estimator.estimateEqual(val, total_rows);
     }
-    else if (op == "less" || op == "lessThan")
+    else if (op == "less" || op == "lessOrEquals")
     {
         if (dummy)
-            return default_normal_cond_factor;
-        return estimator.estimateLess(val) / total_count;
+            return default_normal_cond_factor * total_rows;
+        return estimator.estimateLess(val, total_rows);
     }
-    else if (op == "greater" || op == "greaterThan")
+    else if (op == "greater" || op == "greaterOrEquals")
     {
         if (dummy)
-            return default_normal_cond_factor;
-        return estimator.estimateGreater(val) / total_count;
+            return default_normal_cond_factor * total_rows;
+        return estimator.estimateGreater(val, total_rows);
     }
     else
-        return default_unknown_cond_factor;
+        return default_unknown_cond_factor * total_rows;
 }
 
+void ConditionSelectivityEstimator::merge(String part_name, UInt64 part_rows, ColumnStatisticsPtr column_stat)
+{
+    if (!part_names.contains(part_name))
+    {
+        total_rows += part_rows;
+        part_names.insert(part_name);
+    }
+    if (column_stat != nullptr)
```

Comment:
Assuming that `column_stat` is nullptr (e.g. because the statistics were not materialized): We unconditionally update `total_rows` (l. 194). Will that mess with selectivity estimations (as `total_rows` will encompass all parts but `column_estimators` only some parts)?

Reply:
the selectivity only sum the parts which have statistics, and when we call `estimateRowCount`, we return `total_rows` * selectivity.

So actually this is a RowCountEstimator rather than Selectivity Estimator